File: frontend/src/services/suggestions.ts
// Copyright (c) 2024-2026 Pikar AI. All rights reserved.
// Proprietary and confidential. See LICENSE file for details.

/**
 * Frontend API client for the suggestion chip endpoint.
 *
 * Fetches persona-aware, time-of-day-sensitive suggestion chips
 * from GET /suggestions with a 30-second in-memory cache.
 */

import { fetchWithAuth } from '@/services/api';

export interface SuggestionItem {
  text: string;
  category: string;
}

// ---------------------------------------------------------------------------
// Simple in-memory cache (30 seconds)
// ---------------------------------------------------------------------------

let _cachedResult: SuggestionItem[] | null = null;
let _cachedAt = 0;
let _cachedKey = '';
const CACHE_TTL_MS = 30_000;

/**
 * Fetch personalized suggestion chips from the backend.
 *
 * Results are cached for 30 seconds keyed by persona to avoid
 * refetching on every render cycle.
 */
// ---------------------------------------------------------------------------
// Workflow discovery types
// ---------------------------------------------------------------------------

export interface WorkflowMatch {
  name: string;
  description: string;
  category: string;
  match_score: number;
}

export interface ContentTemplate {
  name: string;
  description: string;
  category: string;
  icon: string;
  example_prompt: string;
}

// ---------------------------------------------------------------------------
// Workflow discovery API
// ---------------------------------------------------------------------------

/**
 * Search workflows by natural-language intent.
 *
 * Calls GET /suggestions/workflows?query=... and returns scored matches.
 */
export async function searchWorkflows(query: string): Promise<WorkflowMatch[]> {
  const response = await fetchWithAuth(
    `/suggestions/workflows?query=${encodeURIComponent(query)}`,
  );
  return response.json();
}

/**
 * Fetch browsable content templates, optionally filtered by category.
 */
export async function fetchContentTemplates(
  category?: string,
): Promise<ContentTemplate[]> {
  const params = category
    ? `?category=${encodeURIComponent(category)}`
    : '';
  const response = await fetchWithAuth(`/suggestions/templates${params}`);
  return response.json();
}

// ---------------------------------------------------------------------------
// Suggestion chips API
// ---------------------------------------------------------------------------

export async function fetchSuggestions(persona: string): Promise<SuggestionItem[]> {
  const now = Date.now();
  const cacheKey = persona;

  if (_cachedResult && _cachedKey === cacheKey && now - _cachedAt < CACHE_TTL_MS) {
    return _cachedResult;
  }

  const hour = new Date().getHours();
  const response = await fetchWithAuth(`/suggestions?persona=${encodeURIComponent(persona)}&hour=${hour}`);
  const data: SuggestionItem[] = await response.json();

  _cachedResult = data;
  _cachedAt = now;
  _cachedKey = cacheKey;

  return data;
}

File: app/routers/suggestions.py
# Copyright (c) 2024-2026 Pikar AI. All rights reserved.
# Proprietary and confidential. See LICENSE file for details.

"""Suggestion chip endpoint for persona-aware chat prompts.

Returns personalized, time-of-day-sensitive suggestion chips
for the chat interface.
"""

from __future__ import annotations

import logging
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, Query

from app.routers.onboarding import get_current_user_id
from app.services.suggestion_service import SuggestionItem, get_suggestions
from app.services.workflow_discovery_service import (
    ContentTemplate,
    WorkflowMatch,
    get_content_templates,
    search_workflows_by_intent,
)

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/suggestions", tags=["Suggestions"])


@router.get("", response_model=list[SuggestionItem])
async def suggestions_endpoint(
    persona: str = Query(..., description="User persona key"),
    hour: int | None = Query(
        None,
        description="Hour of day (0-23), defaults to server UTC hour",
    ),
    recent_activity: str | None = Query(
        None,
        description="Comma-separated recent activity keys",
    ),
    _user_id: str = Depends(get_current_user_id),
) -> list[SuggestionItem]:
    """Return 4-6 personalized suggestion chips.

    Suggestions are tailored to the user's persona, time of day,
    and recent activity history.
    """
    effective_hour = hour if hour is not None else datetime.now(tz=timezone.utc).hour
    activity_list = (
        [a.strip() for a in recent_activity.split(",") if a.strip()]
        if recent_activity
        else None
    )

    return await get_suggestions(
        persona=persona,
        hour=effective_hour,
        recent_activity=activity_list,
    )


@router.get("/workflows", response_model=list[WorkflowMatch])
async def workflow_search_endpoint(
    query: str = Query(..., description="Natural-language description of desired workflow"),
    _user_id: str = Depends(get_current_user_id),
) -> list[WorkflowMatch]:
    """Search workflows by natural-language intent.

    Returns up to 5 scored workflow matches based on keyword
    and substring overlap with available workflow templates.
    """
    return await search_workflows_by_intent(query)


@router.get("/templates", response_model=list[ContentTemplate])
async def content_templates_endpoint(
    category: str | None = Query(
        None,
        description="Filter templates by category",
    ),
    _user_id: str = Depends(get_current_user_id),
) -> list[ContentTemplate]:
    """Return browsable content templates, optionally filtered by category."""
    return await get_content_templates(category=category)
